warning: glib/Android.mk has type 100755, expected 100644
diff --git a/glib/Android.mk b/glib/Android.mk
index 764f07c087c30e176a929f518f1509fe1a734d37..640a5368b195f64a326051de77086a6f3b9db6a3 100755
--- a/glib/Android.mk
+++ b/glib/Android.mk
@@ -45,5 +45,12 @@ LOCAL_CFLAGS:= \
 	-DANDROID_STUB
 
 LOCAL_MODULE:=libglib
+LOCAL_MODULE_TAGS := optional
+LOCAL_VENDOR_MODULE := true
+LOCAL_MULTILIB := 64
+
+LOCAL_EXPORT_C_INCLUDE_DIRS := \
+	$(LOCAL_PATH)/../ \
+	$(LOCAL_PATH)
 
 include $(BUILD_SHARED_LIBRARY)
