diff --git a/glib/gconvert.c b/glib/gconvert.c
--- a/glib/gconvert.c
+++ b/glib/gconvert.c
@@ -89,2 +89,2 @@ try_conversion (const char *to_codeset,
-static gboolean
+static gboolean G_GNUC_UNUSED
 try_to_aliases (const char **to_aliases,
diff --git a/glib/gmain.c b/glib/gmain.c
--- a/glib/gmain.c
+++ b/glib/gmain.c
@@ -3665,11 +3665,7 @@ g_child_watch_source_init_single (void)
   sigaction (SIGCHLD, &action, NULL);
 }
 
-#ifndef ANDROID_STUB
 G_GNUC_NORETURN static gpointer
-#else
-void
-#endif
 child_watch_helper_thread (gpointer data) 
 {
   while (1)
diff --git a/glib/gutf8.c b/glib/gutf8.c
--- a/glib/gutf8.c
+++ b/glib/gutf8.c
@@ -397,6 +397,7 @@ g_utf8_strncpy (gchar       *dest,
   return dest;
 }
 
+#ifndef ANDROID_STUB
 G_LOCK_DEFINE_STATIC (aliases);
 
 static GHashTable *
@@ -446,7 +447,6 @@ get_alias_hash (void)
 /* As an abuse of the alias table, the following routines gets
  * the charsets that are aliases for the canonical name.
  */
-#ifndef ANDROID_STUB
 G_GNUC_INTERNAL const char ** 
 _g_charset_get_aliases (const char *canonical_name)
 {
@@ -460,2 +460,2 @@ _g_charset_get_aliases (const char *canonical_name)
-static gboolean
+static gboolean G_GNUC_UNUSED
 g_utf8_get_charset_internal (const char  *raw_data,
