diff --git a/services/core/java/com/android/server/StorageManagerService.java b/services/core/java/com/android/server/StorageManagerService.java
index 99c59fdb550..c79a38c8061 100644
--- a/services/core/java/com/android/server/StorageManagerService.java
+++ b/services/core/java/com/android/server/StorageManagerService.java
@@ -856,7 +856,8 @@ class StorageManagerService extends IStorageManager.Stub
         } else if (remote == 1) {
             res = true;
         } else {
-            res = true;
+            res = SystemProperties.getBoolean(
+                    "ro.sys.isolated_storage", true);
         }
 
         Slog.d(TAG, "Isolated storage local flag " + local + " and remote flag "
